### ISSUE TITLE
OutlierDetection decrements ejectTimeBackoff too quickly

### ISSUE THREAD

**Issue by lobkovilya** [envoyproxy/envoy#21142]
*Title*: OutlierDetection decrements `ejectTimeBackoff` too quickly

*Description*:

On every tick of the `intervalTimer` value of `ejectTimeBackoff` will be decremented for all not ejected hosts.
```c++
if (!host->healthFlagGet(Host::HealthFlag::FAILED_OUTLIER_CHECK)) {
  // Node seems to be healthy and was not ejected since the last check.
  if (monitor->ejectTimeBackoff() != 0) {
    monitor->ejectTimeBackoff()--;
  }
  return;
}
```
But during the `interval` there could be no evidence that host actually became healthy. 

*Repro steps*:

Let's say we have a host that constantly returns 500 status codes and we configured detector for 3 consecutive 5xx and the interval is equal to 1s:

1. Send 3 requests to the host, each request returns 500
2. OutlierDetection ejects this host for `baseEjectionTime * ejectTimeBackoff` where `ejectTimeBackoff` is equal 1
3. `ejectTimeBackoff++`
4. OutlierDetection unejects host after `baseEjectionTime`
5. After 1s `ejectTimeBackoff` is decremented
6. Send 3 requests to the host again, each returns 500
7. OutlierDetection ejects this host for `baseEjectionTime * ejectTimeBackoff` where `ejectTimeBackoff` is equal 1 again

Expected behaviour: step 7 should eject the host for `baseEjectionTime * 2`.
So at the step 5 OutlierDetection decrements `ejectTimeBackoff` without any knowledge if host recovered or still unhealthy.

*Conclusion*

This behaviour is confusing and not really documented properly. I think `ejectTimeBackoff` should be decremented only when there is convincing evidence that host is recovered.  

xref: https://github.com/envoyproxy/envoy/pull/14235

**Comment by lizan** [comment]
@cpakulski @snowp

**Comment by mattklein123** [comment]
I agree this could potentially be improved, though I'm not sure the extra complexity is worth it. I would recommend starting with a doc PR to clarify the current behavior. Thank you!

**Comment by cpakulski** [comment]
I think that the simplest solution would be to introduce another state called for example Host::HealthFlag::POSSIBLY_HEALTHY which would be equivalent to Host::HealthFlag::HEALTHY except for specific logic like backoff time management. That state would last until the first non-error response.

I was thinking about improving docs, but it is not simple either without describing internal state machine. If the interval is long, there is a higher chance that a node will be ejected again, before backoff counter is decremented. Things are not deterministic here because the final state depends on timing of traffic and calling internal timer.
